### PR TITLE
Add options along with tests for the function-name rule

### PR DESCRIPTION
**Pull request by ckhatri**
Related to issue #498.

If y'all have any suggestions, let me know! 

Edit: Also auto fixed all the " to ' for the linter in FunctionNameRuleTests.ts

### CODE REVIEW

File: src/functionNameRule.ts [hunk 1]
```diff
@@ -11,6 +11,36 @@ const PROTECTED_METHOD_REGEX = 'protected-method-regex';
 const STATIC_METHOD_REGEX = 'static-method-regex';
 const FUNCTION_REGEX = 'function-regex';
 
+const VALIDATE_PRIVATE_STATICS_AS_PRIVATE = 'validate-private-statics-as-private';
+const VALIDATE_PRIVATE_STATICS_AS_STATIC = 'validate-private-statics-as-static';
+const VALIDATE_PRIVATE_STATICS_AS_EITHER = 'validate-private-statics-as-either';
+const NOT_VALID_ARG = 'none';
```
Comment (JoshuaKGoldberg) [inline src/functionNameRule.ts]: Pretending there isn't a history of discussion in the issue, I'm getting a little confused by these options. If someone just passes `[true]`, what are private statics validated as? How is that different from explicitly passing the `-either` option?

Comment (ckhatri) [reply src/functionNameRule.ts]: Oh didn't know, got that updated!

File: src/functionNameRule.ts [hunk 2]
```diff
@@ -20,8 +50,31 @@ export class Rule extends Lint.Rules.AbstractRule {
         ruleName: 'function-name',
         type: 'maintainability',
         description: 'Applies a naming convention to function names and method names',
-        options: null,
-        optionsDescription: '',
+        optionsDescription: Lint.Utils.dedent`
+            Function styles should be consistent throughout the code.
+            Users may want functions with multiple descriptors to be validated a certain way.
+            An optional argument specifies validation for private static methods:
+            * \`${VALIDATE_PRIVATE_STATICS_AS_PRIVATE.toString()}\` enforces validation as private.
```
[the remark this answers is not in the document]
Comment (ckhatri) [reply src/functionNameRule.ts]: Got that fixed!

**Comment by ckhatri** [comment]
Not sure why the build is failing in Travis, seems to pass locally. Sorry!

**Comment by JoshuaKGoldberg** [comment]
> build failing

Huh, that's fun. Looks like the build is broken for TypeScript 3.1, which was just released today. That's definitely not an issue on your end! I'll file a separate issue & resolve it in another PR.

**Comment by JoshuaKGoldberg** [comment]
Looks great, thanks!

I can't restart the AppVeyor build, but Travis is passing so this seems perfect.

**Pull request merged**